{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "26f31433",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "list_annot_run1\n",
      "df_Geneset_LOF1: \n",
      "\n",
      "          GENE                       SNP\n",
      "0        PERM1              1:979654:C:T\n",
      "1       TTLL10            1:1180842:C:CG\n",
      "2     C1QTNF12            1:1246564:T:TG\n",
      "3       SCNN1D             1:1284002:C:T\n",
      "4       ATAD3B  1:1482288:T:TCTTGGAGTCCC\n",
      "...        ...                       ...\n",
      "1927    MAGEC3           X:141881464:C:T\n",
      "1928   SPANXN2          X:143720597:AT:A\n",
      "1929   MAGEA10           X:152135687:T:C\n",
      "1930    PNMA6F           X:153320578:C:A\n",
      "1931    WASH6P           X:156021693:G:A\n",
      "\n",
      "[1932 rows x 2 columns]\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import sys\n",
    "import argparse\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "from collections import defaultdict\n",
    "import json\n",
    "import subprocess\n",
    "import scipy\n",
    "from collections import Counter\n",
    "from io import StringIO\n",
    "from scipy.stats import chi2_contingency\n",
    "\n",
    "\n",
    "file_vcf=\"./step1/SIOP_FAME_annotated.101524.vcf\"\n",
    "\n",
    "list_annot_run=\"list_annot_run1\"\n",
    "\n",
    "print(list_annot_run)\n",
    "#list_annot_run1\n",
    "\n",
    "GT_control='./step1/output4_GT_control.tsv'\n",
    "GT_case='./step1/output4_GT_case.tsv'\n",
    "\n",
    "list_annot_directory='./step2_to_4/' #list_LOF1.tsv, list_LOF2.tsv ...so on\n",
    "suffix='.tsv'\n",
    "\n",
    "geneset_LoF1='./step2_to_4/study_LoF1.SetID'\n",
    "geneset_LoF2='./step2_to_4/study_LoF2.SetID'\n",
    "geneset_missense='./step2_to_4/study_missense.SetID'\n",
    "geneset_moderate='./step2_to_4/study_moderate.SetID'\n",
    "geneset_modifier='./step2_to_4/study_modifier.SetID'\n",
    "geneset_low='./step2_to_4/study_low.SetID'\n",
    "geneset_synonymous='./step2_to_4/study_synonymous.SetID'\n",
    "\n",
    "list_annot_LoF1=['LoF1']\n",
    "list_annot_LoF2=['LoF2']\n",
    "list_annot_missense=['missense']\n",
    "list_annot_moderate=['moderate']\n",
    "list_annot_modifier=['modifier']\n",
    "list_annot_low=['low']\n",
    "list_annot_synonymous=['synonymous']\n",
    "\n",
    "list_annot_run1=['LoF1']\n",
    "list_annot_run2=['LoF1','LoF2','missense']\n",
    "list_annot_run3=['LoF1','LoF2','missense','moderate','modifier','low']\n",
    "list_annot_run4=['synonymous']\n",
    "\n",
    "df_geneset_LoF1=pd.read_csv(geneset_LoF1, sep='\\t', header=None)\n",
    "df_geneset_LoF2=pd.read_csv(geneset_LoF2, sep='\\t', header=None)\n",
    "df_geneset_missense=pd.read_csv(geneset_missense, sep='\\t', header=None)\n",
    "df_geneset_moderate=pd.read_csv(geneset_moderate, sep='\\t', header=None)\n",
    "df_geneset_modifier=pd.read_csv(geneset_modifier, sep='\\t', header=None)\n",
    "df_geneset_low=pd.read_csv(geneset_low, sep='\\t', header=None)\n",
    "df_geneset_synonymous=pd.read_csv(geneset_synonymous, sep='\\t', header=None)\n",
    "\n",
    "\n",
    "df_geneset_LoF1.columns=['GENE','SNP']\n",
    "df_geneset_LoF2.columns=['GENE','SNP']\n",
    "df_geneset_missense.columns=['GENE','SNP']\n",
    "df_geneset_moderate.columns=['GENE','SNP']\n",
    "df_geneset_modifier.columns=['GENE','SNP']\n",
    "df_geneset_low.columns=['GENE','SNP']\n",
    "df_geneset_synonymous.columns=['GENE','SNP']\n",
    "print(\"df_Geneset_LOF1: \\n\")\n",
    "print(df_geneset_LoF1)\n",
    "\n",
    "list_geneset_LoF1=df_geneset_LoF1['SNP'].tolist()\n",
    "list_geneset_LoF2=df_geneset_LoF2['SNP'].tolist()\n",
    "list_geneset_missense=df_geneset_missense['SNP'].tolist()\n",
    "list_geneset_moderate=df_geneset_moderate['SNP'].tolist()\n",
    "list_geneset_modifier=df_geneset_modifier['SNP'].tolist()\n",
    "list_geneset_low=df_geneset_low['SNP'].tolist()\n",
    "list_geneset_synonymous=df_geneset_synonymous['SNP'].tolist()\n",
    "\n",
    "dicts_GT_count_geneset={}\n",
    "dicts_GT_count_geneset['LoF1']=list_geneset_LoF1\n",
    "dicts_GT_count_geneset['LoF2']=list_geneset_LoF2\n",
    "dicts_GT_count_geneset['missense']=list_geneset_missense\n",
    "dicts_GT_count_geneset['moderate']=list_geneset_moderate\n",
    "dicts_GT_count_geneset['modifier']=list_geneset_modifier\n",
    "dicts_GT_count_geneset['low']=list_geneset_low\n",
    "dicts_GT_count_geneset['synonymous']=list_geneset_synonymous\n",
    "\n",
    "#list_annot_run=list_annot_run2\n",
    "#list_annot_run=[list_annot_run2, list_annot_run3, list_annot_run4]\n",
    "\n",
    "#Define different categories\n",
    "list_LoF1=['stop_gained','splice_acceptor_variant','splice_donor_variant','frameshift_variant','transcript_ablation']\n",
    "list_LoF2=['stop_lost','start_lost','transcript_amplification','feature_elongation','feature_truncation']\n",
    "list_missense=['missense_variant']\n",
    "list_moderate=['inframe_insertion','inframe_deletion','protein_altering_variant']\n",
    "list_modifier=['NMD_transcript_variant','3_prime_UTR_variant','5_prime_UTR_variant','TF_binding_site_variant','intergenic_variant','intron_variant','mature_miRNA_variant','non_coding_transcript_exon_variant','non_coding_transcript_variant','downstream_gene_variant','regulatory_region_variant','upstream_gene_variant','coding_sequence_variant','TFBS_ablation','TFBS_amplification','regulatory_region_ablation','regulatory_region_amplification','intergenic_variant','sequence_variant']\n",
    "list_low=['splice_donor_5th_base_variant','splice_region_variant','splice_donor_region_variant','splice_polypyrimidine_tract_variant','incomplete_terminal_codon_variant']\n",
    "list_synonymous=['synonymous_variant','start_retained_variant','stop_retained_variant']\n",
    "dicts_annotation={'MANE':0,'LoF1':1,'LoF2':2,'missense':3,'moderate':4,'modifier':5,'low':6,'synonymous':7}\n",
    "\n",
    "list_LoF1=['MANE_' + x for x in list_LoF1 if str(x)] + list_LoF1\n",
    "list_LoF2=['MANE_' + x for x in list_LoF2 if str(x)] + list_LoF2\n",
    "list_missense=['MANE_' + x for x in list_missense if str(x)] + list_missense\n",
    "list_moderate=['MANE_' + x for x in list_moderate if str(x)] + list_moderate\n",
    "list_modifier=['MANE_' + x for x in list_modifier if str(x)] + list_modifier\n",
    "list_low=['MANE_' + x for x in list_low if str(x)] + list_low\n",
    "list_synonymous=['MANE_' + x for x in list_synonymous if str(x)] + list_synonymous\n",
    "\n",
    "def nested_dict(n, type):\n",
    "    if n == 1:\n",
    "        return defaultdict(type)\n",
    "    else:\n",
    "        return defaultdict(lambda: nested_dict(n-1, type))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63993453",
   "metadata": {},
   "source": [
    "## Manually checking counts for controls for individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "008a1840",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12:57625689:T:G\n",
      "['1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '0/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '0/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '0/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '0/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '0/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1']\n",
      "Final index [64, 218, 245, 286, 314]\n",
      "individual ['QF00233_0001', 'QG01076_0001', 'QG01244_0001', 'QG01550_0001', 'QG01757_0001']\n"
     ]
    }
   ],
   "source": [
    "with open(GT_control) as inny:\n",
    "    list_control_GT=[]\n",
    "    list_control_nonMiss=[]\n",
    "    list_ID=[]\n",
    "    list_individual=[]\n",
    "    \n",
    "    for line in inny:\n",
    "        temp_id=line.rstrip().split('\\t')[0]\n",
    "        temp_GT=line.rstrip().split('\\t')[1:]\n",
    "        \n",
    "        if temp_id=='ID':\n",
    "            list_samples=temp_GT\n",
    "        else:#updated the following code so that it counts all indices/individuals with deleterious mutation. Before the code was only counting one individual\n",
    "            if temp_id == '12:57625689:T:G': #1:152312600:CACTG:C for FLG gene\n",
    "                print(\"12:57625689:T:G\")\n",
    "                print(temp_GT)\n",
    " \n",
    "            if temp_GT.count('0/0') < temp_GT.count('1/1'): #Make sure you are counting for minor allele\n",
    "                # Find all indices for '0/1', '1/0', '1/1'\n",
    "                index = [i for i, gt in enumerate(temp_GT) if gt in ['0/1', '1/0', '0/0']]\n",
    "            elif temp_GT.count('0/0') >= temp_GT.count('1/1'):\n",
    "                index = [i for i, gt in enumerate(temp_GT) if gt in ['0/1', '1/0', '1/1']]\n",
    "                \n",
    "            if temp_id == '12:57625689:T:G':\n",
    "                print(\"Final index\", index)\n",
    "\n",
    "            list_individual_curr = [list_samples[i] for i in index]\n",
    "            list_individual.append(list_individual_curr)\n",
    "            \n",
    "            if temp_id == '12:57625689:T:G':\n",
    "                print(\"individual\", list_individual_curr)\n",
    "            \n",
    "            # Calculating counts\n",
    "            if temp_GT.count('0/0') < temp_GT.count('1/1'):\n",
    "                count_GT=temp_GT.count('0/1')+temp_GT.count('1/0')+temp_GT.count('0/0')*2\n",
    "            elif temp_GT.count('0/0') >= temp_GT.count('1/1'):\n",
    "                count_GT=temp_GT.count('0/1')+temp_GT.count('1/0')+temp_GT.count('1/1')*2\n",
    "\n",
    "            count_nonMiss = 2 * (temp_GT.count('0/1') + temp_GT.count('1/0') + temp_GT.count('1/1') + temp_GT.count('0/0'))\n",
    "            \n",
    "            list_control_GT.append(count_GT)\n",
    "            list_control_nonMiss.append(count_nonMiss)\n",
    "            list_ID.append(temp_id)\n",
    "\n",
    "summary_GT_control['SNP']=list_ID\n",
    "summary_GT_control['CountGT_control']=list_control_GT\n",
    "summary_GT_control['nonMiss_control']=list_control_nonMiss\n",
    "summary_GT_control['individual_control']=list_individual\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d09ed88c",
   "metadata": {},
   "source": [
    "## Manually checking counts for cases for individuals\n",
    "\n",
    "### Using MUC22 gene as example, which has 1 LoF1 mutation in one variant: chr6:31034713:C:A"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e4d4fe7",
   "metadata": {},
   "source": [
    "## Modifying the code so that it spits out corrrect counts for individuals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3d6a9ae1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12:57625689:T:G\n",
      "['1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '0/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '0/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1', '1/1']\n",
      "Final index [158, 170]\n",
      "individual ['QG01412_0001', 'QG01511_0001']\n"
     ]
    }
   ],
   "source": [
    "with open(GT_case) as inny:\n",
    "    list_case_GT = []\n",
    "    list_case_nonMiss = []\n",
    "    list_ID = []\n",
    "    list_individual = []\n",
    "    \n",
    "    for line in inny:\n",
    "        temp_id = line.rstrip().split('\\t')[0]\n",
    "        temp_GT = line.rstrip().split('\\t')[1:]\n",
    "        \n",
    "        if temp_id == 'ID':\n",
    "            list_samples = temp_GT\n",
    "        else: #updated the following code so that it counts all indices/individuals with deleterious mutation. Before the code was only counting one individual not considering the minor allele\n",
    "            if temp_id == '12:57625689:T:G': #1:152312600:CACTG:C for FLG gene\n",
    "                print(\"12:57625689:T:G\")\n",
    "                print(temp_GT)\n",
    "            if temp_GT.count('0/0') < temp_GT.count('1/1'): #Make sure you are counting for minor allele\n",
    "                # Find all indices for '0/1', '1/0', '1/1'\n",
    "                index = [i for i, gt in enumerate(temp_GT) if gt in ['0/1', '1/0', '0/0']] \n",
    "            elif temp_GT.count('0/0') >= temp_GT.count('1/1'):\n",
    "                index = [i for i, gt in enumerate(temp_GT) if gt in ['0/1', '1/0', '1/1']]\n",
    "                \n",
    "            if temp_id == '12:57625689:T:G':\n",
    "                print(\"Final index\", index)\n",
    "  \n",
    "    \n",
    "            list_individual_curr = [list_samples[i] for i in index]\n",
    "            list_individual.append(list_individual_curr)\n",
    "            \n",
    "            if temp_id == '12:57625689:T:G': \n",
    "                print(\"individual\", list_individual_curr)\n",
    "\n",
    "            # Calculating counts\n",
    "            # Calculating counts\n",
    "            if temp_GT.count('0/0') < temp_GT.count('1/1'):\n",
    "                count_GT=temp_GT.count('0/1')+temp_GT.count('1/0')+temp_GT.count('0/0')*2\n",
    "            elif temp_GT.count('0/0') >= temp_GT.count('1/1'):\n",
    "                count_GT=temp_GT.count('0/1')+temp_GT.count('1/0')+temp_GT.count('1/1')*2\n",
    "                \n",
    "            count_nonMiss = 2 * (temp_GT.count('0/1') + temp_GT.count('1/0') + temp_GT.count('1/1') + temp_GT.count('0/0'))\n",
    "            \n",
    "            list_case_GT.append(count_GT)\n",
    "            list_case_nonMiss.append(count_nonMiss)\n",
    "            list_ID.append(temp_id)\n",
    "\n",
    "summary_GT_case['SNP']=list_ID\n",
    "summary_GT_case['CountGT_case']=list_case_GT\n",
    "summary_GT_case['nonMiss_case']=list_case_nonMiss\n",
    "summary_GT_case['individual_case']=list_individual\n",
    "\n",
    "summary_GT_case_control=pd.merge(summary_GT_control,summary_GT_case,on=['SNP'],how='inner')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fcba6676",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SNP</th>\n",
       "      <th>CountGT_control</th>\n",
       "      <th>nonMiss_control</th>\n",
       "      <th>individual_control</th>\n",
       "      <th>CountGT_case</th>\n",
       "      <th>nonMiss_case</th>\n",
       "      <th>individual_case</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>250302</th>\n",
       "      <td>12:57625157:G:T</td>\n",
       "      <td>1</td>\n",
       "      <td>654</td>\n",
       "      <td>[QG01308_0001]</td>\n",
       "      <td>3</td>\n",
       "      <td>410</td>\n",
       "      <td>[QG01409_0001, QG01511_0001, QG01829_0001]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    SNP  CountGT_control  nonMiss_control individual_control  \\\n",
       "250302  12:57625157:G:T                1              654     [QG01308_0001]   \n",
       "\n",
       "        CountGT_case  nonMiss_case                             individual_case  \n",
       "250302             3           410  [QG01409_0001, QG01511_0001, QG01829_0001]  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "summary_GT_case_control[summary_GT_case_control['SNP'] == '12:57625157:G:T']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "10e43bd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import islice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "59b91fa3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1:16666:G:A': defaultdict(<class 'list'>, {'DDX11L1': 'modifier', 'DDX11L2': 'modifier', 'WASH7P': 'modifier', 'MIR6859-1': 'modifier'}), '1:16792:G:A': defaultdict(<class 'list'>, {'DDX11L1': 'modifier', 'DDX11L2': 'modifier', 'WASH7P': 'modifier', 'MIR6859-1': 'modifier'}), '1:16825:C:A': defaultdict(<class 'list'>, {'DDX11L1': 'modifier', 'DDX11L2': 'modifier', 'WASH7P': 'modifier', 'MIR6859-1': 'modifier'}), '1:17036:T:C': defaultdict(<class 'list'>, {'DDX11L1': 'modifier', 'DDX11L2': 'modifier', 'WASH7P': 'modifier', 'MIR6859-1': 'modifier'}), '1:17181:G:C': defaultdict(<class 'list'>, {'DDX11L1': 'modifier', 'DDX11L2': 'modifier', 'WASH7P': 'modifier', 'MIR6859-1': 'modifier'})}\n"
     ]
    }
   ],
   "source": [
    "\n",
    "dicts=nested_dict(2,list)\n",
    "with open(file_vcf) as vcf_in:\n",
    "    for line in vcf_in:\n",
    "        if line[0]=='#':\n",
    "            continue\n",
    "        tmp_Variation,tmp_Location,tmp_Allele,tmp_Gene,tmp_Feature,tmp_Feature_type,tmp_Consequence,tmp_cDNA_position,tmp_CDS_position,tmp_Protein_position,tmp_Amino_acids,tmp_Codons,tmp_Existing_variation,tmp_Extra = line.rstrip().split('\\t')\n",
    "        tmp_annot=tmp_Consequence\n",
    "        if 'SYMBOL=' in tmp_Extra:\n",
    "            result=re.search('SYMBOL=(.*);SYMBOL_SOURCE', tmp_Extra)\n",
    "            tmp_gene=result.group(1)\n",
    "        #if more than one annotation,should be assign the most deleterious one\n",
    "        list_annotation=tmp_annot.split(',')\n",
    "        if 'MANE' not in tmp_Extra:\n",
    "            if len(list_annotation) == 1:\n",
    "                dicts[tmp_Variation][tmp_gene].append(list_annotation[0])\n",
    "            elif len(list_annotation) > 1:\n",
    "                for annot_in_list in list_annotation:\n",
    "                    dicts[tmp_Variation][tmp_gene].append(annot_in_list)\n",
    "        elif 'MANE' in tmp_Extra:\n",
    "            if len(list_annotation) == 1:\n",
    "                dicts[tmp_Variation][tmp_gene].append('MANE_'+list_annotation[0])\n",
    "            elif len(list_annotation) > 1:\n",
    "                for annot_in_list in list_annotation:\n",
    "                    dicts[tmp_Variation][tmp_gene].append('MANE_'+annot_in_list)\n",
    "\n",
    "for key, val in dicts.items():\n",
    "    for val_key, val_val in val.items():\n",
    "        list_tmp=[]\n",
    "        flag_mane=0\n",
    "        for element_annotation in val_val:\n",
    "            if 'MANE' in element_annotation and flag_mane==0:\n",
    "                list_tmp=[]\n",
    "                list_tmp.append(element_annotation)\n",
    "                flag_mane=1\n",
    "            elif 'MANE' in element_annotation and flag_mane==1:\n",
    "                list_tmp.append(element_annotation)\n",
    "                flag_mane=1\n",
    "            elif 'MANE' not in element_annotation and flag_mane==0:\n",
    "                list_tmp.append(element_annotation)\n",
    "        dicts[key][val_key]=list_tmp\n",
    "\n",
    "for key, val in dicts.items():\n",
    "    for val_key, val_val in val.items():\n",
    "        tmp_deleterious=''\n",
    "        for element_annotation in val_val:\n",
    "            if element_annotation in list_LoF1:\n",
    "                element_annotation_category = 'LoF1'\n",
    "            elif element_annotation in list_LoF2:\n",
    "                element_annotation_category = 'LoF2'\n",
    "            elif element_annotation in list_missense:\n",
    "                element_annotation_category = 'missense'\n",
    "            elif element_annotation in list_moderate:\n",
    "                element_annotation_category = 'moderate'\n",
    "            elif element_annotation in list_modifier:\n",
    "                element_annotation_category = 'modifier'\n",
    "            elif element_annotation in list_low:\n",
    "                element_annotation_category = 'low'\n",
    "            elif element_annotation in list_synonymous:\n",
    "                element_annotation_category = 'synonymous'\n",
    "            if tmp_deleterious == '':\n",
    "                tmp_deleterious=element_annotation_category\n",
    "            elif tmp_deleterious != '':\n",
    "                if dicts_annotation[element_annotation_category] <= dicts_annotation[tmp_deleterious]:\n",
    "                    tmp_deleterious=element_annotation_category\n",
    "                else:\n",
    "                    continue\n",
    "        dicts[key][val_key]=tmp_deleterious\n",
    "print(dict(islice(dicts.items(), 5)))\n",
    "\n",
    "#returns variant for the gene and annoation(LOF1,LOF2...) using the dictionary\n",
    "def get_key(gene, annot):\n",
    "    list_key=[]\n",
    "    for key, value in dicts.items():\n",
    "        for value_key, value_value in value.items():\n",
    "            if value_key == gene and value_value in annot:\n",
    "                list_key.append(key)\n",
    "    return list_key\n",
    "\n",
    "#list_annot_run=list_annot_run2\n",
    "#list_annot_run=list_annot_run2\n",
    "dicts_skato={}\n",
    "list_gene=[]\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "6add615a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12:57625157:G:T: SLC26A10P: LoF1\n",
      "12:57625689:T:G: SLC26A10P: LoF1\n"
     ]
    }
   ],
   "source": [
    "# Extract and print only the SNP key and \"FASTKD1: missense\" pairs\n",
    "for snp, genes in dicts.items():\n",
    "    if 'SLC26A10P' in genes and genes['SLC26A10P'] == 'LoF1':\n",
    "        print(f\"{snp}: SLC26A10P: {genes['SLC26A10P']}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "1f5f8f05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'1:152302520:C:G': defaultdict(<class 'list'>, {'FLG': 'modifier', 'FLG-AS1': 'modifier'}), '1:152302521:G:A': defaultdict(<class 'list'>, {'FLG': 'modifier', 'FLG-AS1': 'modifier'}), '1:152302797:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152302977:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152303047:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152303168:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152303212:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152303333:C:T': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152303400:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152303401:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152303407:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152303688:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152303813:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152303976:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152304008:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152304399:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152304449:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152304456:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152304500:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152304608:C:T': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152304624:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152304644:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152304661:G:A': defaultdict(<class 'list'>, {'FLG': 'LoF1', 'FLG-AS1': 'modifier'}), '1:152304661:G:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152304772:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152304831:G:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152304833:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152305048:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305078:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305099:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305139:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152305146:G:T': defaultdict(<class 'list'>, {'FLG': 'LoF1', 'FLG-AS1': 'modifier'}), '1:152305186:A:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305214:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152305222:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305321:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305344:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305366:T:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305602:T:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305624:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305633:C:A': defaultdict(<class 'list'>, {'FLG': 'LoF1', 'FLG-AS1': 'modifier'}), '1:152305737:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152305788:G:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306051:G:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306118:A:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306139:T:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306148:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306194:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306230:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306289:T:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306318:C:T': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152306323:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306351:T:C': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152306353:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306354:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306357:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152306367:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306368:T:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306516:C:T': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152306582:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306589:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152306699:C:T': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152307013:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307085:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307210:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307268:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307311:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152307371:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152307416:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152307420:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307434:T:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307446:T:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307499:G:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307524:C:T': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152307589:T:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307598:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307607:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307685:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307726:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152307728:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152308043:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152308099:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152308130:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152308234:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152308263:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152308661:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152308668:T:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152308777:G:A': defaultdict(<class 'list'>, {'FLG': 'LoF1', 'FLG-AS1': 'modifier'}), '1:152308813:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152308947:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152309003:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152309015:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152309080:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152309164:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152309169:G:T': defaultdict(<class 'list'>, {'FLG': 'LoF1', 'FLG-AS1': 'modifier'}), '1:152309183:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152309248:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152309438:G:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152309549:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152309735:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152309736:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152309920:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152309985:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152309996:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152310022:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152310353:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152310577:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152310664:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152310791:T:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152310837:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152310914:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152310929:G:A': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152310990:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311054:T:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311057:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311277:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311299:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311332:A:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311446:T:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311829:C:T': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152311950:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311974:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311980:T:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152311981:T:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312082:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312173:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312306:C:T': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152312472:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312485:T:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312520:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312526:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312600:CACTG:C': defaultdict(<class 'list'>, {'FLG': 'LoF1', 'FLG-AS1': 'modifier'}), '1:152312603:T:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152312606:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312744:G:C': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152312745:A:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312747:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312813:T:C': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152312814:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312851:T:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312876:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152312893:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152312894:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152313065:T:C': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152313313:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152313331:G:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152313355:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152313368:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152313392:C:G': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152313401:C:T': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152313445:C:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152313446:A:G': defaultdict(<class 'list'>, {'FLG': 'synonymous', 'FLG-AS1': 'modifier'}), '1:152313449:G:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152313481:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152313631:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152314602:G:A': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152314737:T:C': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'}), '1:152314837:C:G': defaultdict(<class 'list'>, {'FLG': 'modifier', 'FLG-AS1': 'modifier'}), '1:152315381:C:T': defaultdict(<class 'list'>, {'FLG': 'missense', 'FLG-AS1': 'modifier'})}\n"
     ]
    }
   ],
   "source": [
    "result = {snp: genes for snp, genes in dicts.items() if 'FLG' in genes}\n",
    "print(result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "771412b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       CHROM    GENPOS                       ID ALLELE0          ALLELE1  \\\n",
      "0          1  33306701  A3GALT2.Mask1.singleton     ref  Mask1.singleton   \n",
      "1          1  33306701       A3GALT2.Mask1.0.01     ref       Mask1.0.01   \n",
      "2          1  33306701  A3GALT2.Mask2.singleton     ref  Mask2.singleton   \n",
      "3          1  33306701       A3GALT2.Mask2.0.01     ref       Mask2.0.01   \n",
      "4          1  33306701  A3GALT2.Mask3.singleton     ref  Mask3.singleton   \n",
      "...      ...       ...                      ...     ...              ...   \n",
      "85523     23  57591997          ZXDB.Mask2.0.01     ref       Mask2.0.01   \n",
      "85524     23  57591997     ZXDB.Mask3.singleton     ref  Mask3.singleton   \n",
      "85525     23  57591997          ZXDB.Mask3.0.01     ref       Mask3.0.01   \n",
      "85526     23  57591997     ZXDB.Mask4.singleton     ref  Mask4.singleton   \n",
      "85527     23  57591997          ZXDB.Mask4.0.01     ref       Mask4.0.01   \n",
      "\n",
      "         A1FREQ    N TEST      BETA        SE     CHISQ    LOG10P  \n",
      "0      0.001883  531  ADD -0.583552  2.034800  0.082246  0.111104  \n",
      "1      0.001883  531  ADD -0.583552  2.034800  0.082246  0.111104  \n",
      "2      0.008459  532  ADD -0.977079  0.810597  1.452950  0.641959  \n",
      "3      0.027256  532  ADD -0.244023  0.450171  0.293838  0.230791  \n",
      "4      0.012218  532  ADD -0.490758  0.677497  0.524711  0.328976  \n",
      "...         ...  ...  ...       ...       ...       ...       ...  \n",
      "85523  0.000943  530  ADD  1.300580  2.506380  0.269268  0.219090  \n",
      "85524  0.001880  532  ADD  2.826000  1.713530  2.719950  1.003920  \n",
      "85525  0.004699  532  ADD  2.205620  0.897031  6.045720  1.855740  \n",
      "85526  0.000943  530  ADD -1.234480  2.688300  0.210869  0.189709  \n",
      "85527  0.006579  532  ADD -0.235895  0.706181  0.111585  0.131739  \n",
      "\n",
      "[85528 rows x 12 columns]\n"
     ]
    }
   ],
   "source": [
    "df_regenie=pd.read_csv(\"./step7/SIOP_FAME.ALL.101524_responder.regenie\", skiprows=1,sep='\\s+')\n",
    "#print(\"regenie summary statistics .tsv file from step9: \\n\")\n",
    "print(df_regenie)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "13584f2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "list_gene=[]\n",
    "for row in df_regenie.itertuples():\n",
    "    tmp_row=getattr(row, 'ID')\n",
    "    tmp_gene=tmp_row.split('.')[0]\n",
    "    list_gene.append(tmp_gene)\n",
    "#print(list_gene)\n",
    "df_regenie['GENE']=list_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "669b15a4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "RUNNING run1\n",
      "regenie summary statistics for selected Mask: \n",
      "\n",
      "       CHROM     GENPOS                  ID ALLELE0     ALLELE1    A1FREQ  \\\n",
      "1          1   33306701  A3GALT2.Mask1.0.01     ref  Mask1.0.01  0.001883   \n",
      "178        1  246840884   AHCTF1.Mask1.0.01     ref  Mask1.0.01  0.007547   \n",
      "192        1  159062694     AIM2.Mask1.0.01     ref  Mask1.0.01  0.001880   \n",
      "271        1  114673041    AMPD1.Mask1.0.01     ref  Mask1.0.01  0.001880   \n",
      "284        1  103617292    AMY2A.Mask1.0.01     ref  Mask1.0.01  0.000945   \n",
      "...      ...        ...                 ...     ...         ...       ...   \n",
      "85107     23  125320027   TEX13C.Mask1.0.01     ref  Mask1.0.01  0.001880   \n",
      "85111     23  124332927   TEX13D.Mask1.0.01     ref  Mask1.0.01  0.001887   \n",
      "85203     23  101010678   TRMT2B.Mask1.0.01     ref  Mask1.0.01  0.002820   \n",
      "85347     23  156021693   WASH6P.Mask1.0.01     ref  Mask1.0.01  0.000951   \n",
      "85436     23   74304311  ZCCHC13.Mask1.0.01     ref  Mask1.0.01  0.005639   \n",
      "\n",
      "         N TEST      BETA        SE     CHISQ    LOG10P     GENE  \n",
      "1      531  ADD -0.583552  2.034800  0.082246  0.111104  A3GALT2  \n",
      "178    530  ADD  0.824283  0.890326  0.857144  0.450336   AHCTF1  \n",
      "192    532  ADD  1.017250  1.572870  0.418279  0.285840     AIM2  \n",
      "271    532  ADD -0.266107  1.732310  0.023597  0.056548    AMPD1  \n",
      "284    529  ADD -1.417580  2.229150  0.404409  0.279989    AMY2A  \n",
      "...    ...  ...       ...       ...       ...       ...      ...  \n",
      "85107  532  ADD -3.508070  1.433020  5.992830  1.842720   TEX13C  \n",
      "85111  530  ADD -0.790485  1.050350  0.566398  0.345156   TEX13D  \n",
      "85203  532  ADD -1.008530  0.976313  1.067090  0.520563   TRMT2B  \n",
      "85347  526  ADD  1.172510  2.937940  0.159275  0.161261   WASH6P  \n",
      "85436  532  ADD -0.465692  0.784272  0.352586  0.257548  ZCCHC13  \n",
      "\n",
      "[1619 rows x 13 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "condition_mask1 = df_regenie['ID'].str.contains('Mask1')\n",
    "condition_mask2 = df_regenie['ID'].str.contains('Mask2')\n",
    "condition_mask3 = df_regenie['ID'].str.contains('Mask3')\n",
    "condition_mask4 = df_regenie['ID'].str.contains('Mask4')\n",
    "condition_af01 = df_regenie['ID'].str.contains('0.01')\n",
    "\n",
    "#print(list_annot_run == \"list_annot_run1\") \n",
    "\n",
    "if list_annot_run == \"list_annot_run1\":\n",
    "    print('RUNNING run1')\n",
    "    combined_condition_af01 = condition_mask1 & condition_af01\n",
    "    df_regenie_Mask_af01 = df_regenie[combined_condition_af01]\n",
    "    list_annot_run=list_annot_run1\n",
    "\n",
    "elif list_annot_run == \"list_annot_run2\":\n",
    "    print('RUNNING run2')\n",
    "    combined_condition_af01 = condition_mask2 & condition_af01\n",
    "    df_regenie_Mask_af01 = df_regenie[combined_condition_af01]\n",
    "    list_annot_run=list_annot_run2\n",
    "\n",
    "elif list_annot_run == \"list_annot_run3\":\n",
    "    print('RUNNING run3')\n",
    "    combined_condition_af01 = condition_mask3 & condition_af01\n",
    "    df_regenie_Mask_af01 = df_regenie[combined_condition_af01]\n",
    "    list_annot_run=list_annot_run3\n",
    "\n",
    "elif list_annot_run == \"list_annot_run4\":\n",
    "    print('RUNNING run4')\n",
    "    combined_condition_af01 = condition_mask4 & condition_af01\n",
    "    df_regenie_Mask_af01 = df_regenie[combined_condition_af01]\n",
    "    list_annot_run=list_annot_run4\n",
    "\n",
    "print(\"regenie summary statistics for selected Mask: \\n\")\n",
    "\n",
    "print(df_regenie_Mask_af01)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "819e249e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "done!\n",
      "dictionary for regenie with selected Mask: \n",
      "\n"
     ]
    }
   ],
   "source": [
    "dicts_skato={}\n",
    "#makes a new dictionary dicts_skato (gene:variant) by taking in the GENE and list_annot_run(LOF1..)\n",
    "for row in df_regenie_Mask_af01.itertuples():\n",
    "    tmp_id=getattr(row, 'GENE')\n",
    "    #tmp_setID, tmp_Mask, tmp_afrq=tmp_id.rstrip().split('.',2)\n",
    "    #print(tmp_setID) \n",
    "    #print(tmp_Mask)\n",
    "    #print(tmp_afrq)\n",
    "    list_gene=[]\n",
    "    list_gene=get_key(tmp_id, list_annot_run)\n",
    "    if tmp_id not in dicts_skato:\n",
    "        dicts_skato[tmp_id]=list_gene\n",
    "    else:\n",
    "        print('extists! error')\n",
    "print(\"done!\")\n",
    "print(\"dictionary for regenie with selected Mask: \\n\")\n",
    "#print(dicts_skato)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "50286c68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "annot:  LoF1\n",
      "SLC26A10P ['12:57625157:G:T', '12:57625689:T:G']\n",
      "total\n",
      "SNP                                      12:57625157:G:T12:57625689:T:G\n",
      "CountGT_control                                                       6\n",
      "nonMiss_control                                                    1308\n",
      "individual_control    [QG01308_0001, QF00233_0001, QG01076_0001, QG0...\n",
      "CountGT_case                                                          5\n",
      "nonMiss_case                                                        820\n",
      "individual_case       [QG01409_0001, QG01511_0001, QG01829_0001, QG0...\n",
      "dtype: object\n",
      "control individuals\n",
      "['QG01308_0001', 'QF00233_0001', 'QG01076_0001', 'QG01244_0001', 'QG01550_0001', 'QG01757_0001']\n",
      "6\n",
      "unique individuals\n",
      "6\n",
      "case individuals\n",
      "['QG01409_0001', 'QG01511_0001', 'QG01829_0001', 'QG01412_0001', 'QG01511_0001']\n",
      "5\n",
      "unique individuals\n",
      "4\n",
      "output_step2: \n",
      "\n",
      "          SNP  CountGT_control_LoF1  CountGT_case_LoF1  nonMiss_control_LoF1  \\\n",
      "0     A3GALT2                     1                  1                  1304   \n",
      "1      AHCTF1                     2                  6                   654   \n",
      "2        AIM2                     1                  1                   654   \n",
      "3       AMPD1                     1                  1                   654   \n",
      "4       AMY2A                     1                  0                   650   \n",
      "...       ...                   ...                ...                   ...   \n",
      "1614   TEX13C                     2                  0                   654   \n",
      "1615   TEX13D                     2                  0                   654   \n",
      "1616   TRMT2B                     3                  0                   654   \n",
      "1617   WASH6P                     0                  1                   646   \n",
      "1618  ZCCHC13                     4                  2                   654   \n",
      "\n",
      "      nonMiss_case_LoF1  #individual_allele_case_LoF1  \\\n",
      "0                   820                             1   \n",
      "1                   406                             6   \n",
      "2                   410                             1   \n",
      "3                   410                             1   \n",
      "4                   408                             0   \n",
      "...                 ...                           ...   \n",
      "1614                410                             0   \n",
      "1615                406                             0   \n",
      "1616                410                             0   \n",
      "1617                406                             1   \n",
      "1618                410                             1   \n",
      "\n",
      "      #individual_allele_control_LoF1  \n",
      "0                                   1  \n",
      "1                                   2  \n",
      "2                                   1  \n",
      "3                                   1  \n",
      "4                                   1  \n",
      "...                               ...  \n",
      "1614                                1  \n",
      "1615                                1  \n",
      "1616                                2  \n",
      "1617                                0  \n",
      "1618                                3  \n",
      "\n",
      "[1619 rows x 7 columns]\n",
      "output_step2_control_GT\n",
      "      CountGT_control_LoF1\n",
      "0                        1\n",
      "1                        2\n",
      "2                        1\n",
      "3                        1\n",
      "4                        1\n",
      "...                    ...\n",
      "1614                     2\n",
      "1615                     2\n",
      "1616                     3\n",
      "1617                     0\n",
      "1618                     4\n",
      "\n",
      "[1619 rows x 1 columns]\n",
      "output_step2_case_GT\n",
      "      CountGT_case_LoF1\n",
      "0                     1\n",
      "1                     6\n",
      "2                     1\n",
      "3                     1\n",
      "4                     0\n",
      "...                 ...\n",
      "1614                  0\n",
      "1615                  0\n",
      "1616                  0\n",
      "1617                  1\n",
      "1618                  2\n",
      "\n",
      "[1619 rows x 1 columns]\n",
      "list type\n",
      "CountGT_case_nonMiss\n",
      "0    820\n",
      "1    406\n",
      "2    410\n",
      "3    410\n",
      "4    408\n",
      "Name: CountGT_case_nonMiss, dtype: int64\n",
      "CountGT_case_All               0\n",
      "CountGT_case_All_other       398\n",
      "CountGT_control_All            0\n",
      "CountGT_control_All_other    636\n",
      "dtype: int64\n",
      "          SNP  CountGT_control_LoF1  CountGT_case_LoF1  nonMiss_control_LoF1  \\\n",
      "0     A3GALT2                     1                  1                  1304   \n",
      "1      AHCTF1                     2                  6                   654   \n",
      "2        AIM2                     1                  1                   654   \n",
      "3       AMPD1                     1                  1                   654   \n",
      "4       AMY2A                     1                  0                   650   \n",
      "...       ...                   ...                ...                   ...   \n",
      "1614   TEX13C                     2                  0                   654   \n",
      "1615   TEX13D                     2                  0                   654   \n",
      "1616   TRMT2B                     3                  0                   654   \n",
      "1617   WASH6P                     0                  1                   646   \n",
      "1618  ZCCHC13                     4                  2                   654   \n",
      "\n",
      "      nonMiss_case_LoF1  #individual_allele_case_LoF1  \\\n",
      "0                   820                             1   \n",
      "1                   406                             6   \n",
      "2                   410                             1   \n",
      "3                   410                             1   \n",
      "4                   408                             0   \n",
      "...                 ...                           ...   \n",
      "1614                410                             0   \n",
      "1615                406                             0   \n",
      "1616                410                             0   \n",
      "1617                406                             1   \n",
      "1618                410                             1   \n",
      "\n",
      "      #individual_allele_control_LoF1  CountGT_control_All  CountGT_case_All  \\\n",
      "0                                   1                    1                 1   \n",
      "1                                   2                    2                 6   \n",
      "2                                   1                    1                 1   \n",
      "3                                   1                    1                 1   \n",
      "4                                   1                    1                 0   \n",
      "...                               ...                  ...               ...   \n",
      "1614                                1                    2                 0   \n",
      "1615                                1                    2                 0   \n",
      "1616                                2                    3                 0   \n",
      "1617                                0                    0                 1   \n",
      "1618                                3                    4                 2   \n",
      "\n",
      "      CountGT_control_nonMiss  CountGT_case_nonMiss  Odds_Ratio  P_Chi-Square  \n",
      "0                        1304                   820    1.590244      1.000000  \n",
      "1                         654                   406    3.799667      0.075355  \n",
      "2                         654                   410    1.595122      1.000000  \n",
      "3                         654                   410    1.595122      1.000000  \n",
      "4                         650                   408    0.794621      1.000000  \n",
      "...                       ...                   ...         ...           ...  \n",
      "1614                      654                   410    0.529603      0.693843  \n",
      "1615                      654                   406    0.534808      0.698505  \n",
      "1616                      654                   410    0.396594      0.435779  \n",
      "1617                      646                   406    3.187192      0.814653  \n",
      "1618                      654                   410    0.955012      1.000000  \n",
      "\n",
      "[1619 rows x 13 columns]\n"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "output_step2=pd.DataFrame()\n",
    "for annot in list_annot_run:\n",
    "    print(\"annot: \",annot)\n",
    "    out_step1='list_'+annot+suffix\n",
    "    #list_annot_directory='/data/Segre_Lab/users/jlama/WES_new.ALL_050824/GeneBurden/FAME/step2/'\n",
    "    out1_step1=os.path.join(list_annot_directory, out_step1)\n",
    "    #print(out1_step1)\n",
    "    df_run=pd.read_csv(out1_step1,sep='\\t')\n",
    "    df_run.columns=['SNP']\n",
    "    df_check_run=pd.DataFrame()\n",
    "    list_check_run=[]\n",
    "    for row in df_run.itertuples():\n",
    "        tmp_snp=getattr(row, 'SNP')\n",
    "        tmp_chr, tmp_pos, tmp_ref, tmp_alt=tmp_snp.split(':')\n",
    "        check_snp=tmp_snp\n",
    "        if check_snp in dicts_GT_count_geneset[annot]:\n",
    "            list_check_run.append(tmp_snp)\n",
    "    df_check_run['SNP']=list_check_run\n",
    "    summary_GT_case_control_run=pd.merge(summary_GT_case_control,df_check_run,on=['SNP'],how='inner')\n",
    "    \n",
    "    new_df=pd.DataFrame()\n",
    "    list_gene=[]\n",
    "    list_case_GT=[]\n",
    "    list_control_GT=[]\n",
    "    list_case_nonMiss=[]\n",
    "    list_control_nonMiss=[]\n",
    "    list_case_individual=[]\n",
    "    list_control_individual=[]\n",
    "    for key, value in dicts_skato.items():\n",
    "        subset=summary_GT_case_control_run.loc[summary_GT_case_control_run[\"SNP\"].isin(value)]\n",
    "        if 'SLC26A10P' in key:\n",
    "            print(key, value)     \n",
    "            #print(subset)\n",
    "        \n",
    "        if subset.empty == False:\n",
    "            #print(\"summary_GT_case_control_run\")\n",
    "            #print(summary_GT_case_control_run)\n",
    "            total=subset.sum()\n",
    "            if 'SLC26A10P' in key:\n",
    "                print(\"total\")\n",
    "                print(total)\n",
    "            \n",
    "            list_gene.append(key)\n",
    "            list_control_GT.append(total[1])\n",
    "            list_control_nonMiss.append(total[2]) #nonMiss_control\n",
    "            list_control_individual.append(len(set(total['individual_control'])))\n",
    "            if 'SLC26A10P' in key:\n",
    "                print(\"control individuals\")\n",
    "                print(total[3])\n",
    "                print(len(total['individual_control']))\n",
    "                print(\"unique individuals\")\n",
    "                print(len(set(total['individual_control'])))\n",
    "            \n",
    "            list_case_individual.append(len(set(total['individual_case'])))\n",
    "            if 'SLC26A10P' in key:\n",
    "                print(\"case individuals\")\n",
    "                print(total[6])\n",
    "                print(len(total['individual_case']))\n",
    "                print(\"unique individuals\")\n",
    "                print(len(set(total['individual_case'])))\n",
    "            list_case_GT.append(total[4])\n",
    "            list_case_nonMiss.append(total[5]) #nonMiss_case\n",
    "            \n",
    "        elif subset.empty == True:\n",
    "            list_gene.append(key)\n",
    "            list_control_GT.append(pd.NA)\n",
    "            list_control_nonMiss.append(pd.NA)\n",
    "            list_control_individual.append(pd.NA)\n",
    "            list_case_individual.append(pd.NA)\n",
    "            list_case_GT.append(pd.NA)\n",
    "            list_case_nonMiss.append(pd.NA)\n",
    "    new_df['SNP']=list_gene\n",
    "    new_df['CountGT_control_'+annot]=list_control_GT\n",
    "    new_df['CountGT_case_'+annot]=list_case_GT\n",
    "    new_df['nonMiss_control_'+annot]=list_control_nonMiss\n",
    "    new_df['nonMiss_case_'+annot]=list_case_nonMiss\n",
    "    new_df['#individual_allele_case_'+annot]=list_case_individual\n",
    "    new_df['#individual_allele_control_'+annot]=list_control_individual\n",
    "    if len(output_step2)==0:\n",
    "        output_step2=new_df\n",
    "    elif len(output_step2)>0:\n",
    "        output_step2=pd.merge(output_step2, new_df, on=['SNP'], how='inner')\n",
    "print(\"output_step2: \\n\")\n",
    "print(output_step2)\n",
    "\n",
    "\n",
    "prefix_control_GT='CountGT_control_'\n",
    "prefix_case_GT='CountGT_case_'\n",
    "prefix_control_nonMiss='nonMiss_control_'\n",
    "prefix_case_nonMiss='nonMiss_case_'\n",
    "output_step2_control_GT=output_step2.filter(like=prefix_control_GT, axis=1)\n",
    "print(\"output_step2_control_GT\")\n",
    "print(output_step2_control_GT)\n",
    "\n",
    "output_step2_case_GT=output_step2.filter(like=prefix_case_GT, axis=1)\n",
    "output_step2_control_nonMiss=output_step2.filter(like=prefix_control_nonMiss, axis=1)\n",
    "output_step2_case_nonMiss=output_step2.filter(like=prefix_case_nonMiss, axis=1)\n",
    "output_step2['CountGT_control_All']=output_step2_control_GT.sum(axis=1, skipna=True)\n",
    "print(\"output_step2_case_GT\")\n",
    "print(output_step2_case_GT)\n",
    "\n",
    "output_step2['CountGT_case_All']=output_step2_case_GT.sum(axis=1, skipna=True)\n",
    "output_step2['CountGT_control_nonMiss']=output_step2_control_nonMiss.sum(axis=1, skipna=True)\n",
    "output_step2['CountGT_case_nonMiss']=output_step2_case_nonMiss.sum(axis=1, skipna=True)\n",
    "\n",
    "print(\"list type\")\n",
    "print(\"CountGT_case_nonMiss\")\n",
    "print(output_step2['CountGT_case_nonMiss'].head())\n",
    "\n",
    "#output_step2['CountGT_case_nonMiss'] = output_step2['CountGT_case_nonMiss'].apply(lambda x: len(x) if isinstance(x, list) else x)\n",
    "#print(output_step2['CountGT_case_nonMiss'].dtype)\n",
    "#print(output_step2['CountGT_case_All'].apply(type).value_counts())\n",
    "#print(output_step2['CountGT_control_All'].apply(type).value_counts())\n",
    "#print(output_step2['CountGT_control_nonMiss'].apply(type).value_counts())\n",
    "#print(output_step2['CountGT_case_nonMiss'].apply(type).value_counts())\n",
    "output_step2['Odds_Ratio']=((output_step2['CountGT_case_All']+1)*(output_step2['CountGT_control_nonMiss']-output_step2['CountGT_control_All']+1))/((output_step2['CountGT_control_All']+1)*(output_step2['CountGT_case_nonMiss']-output_step2['CountGT_case_All']+1))\n",
    "\n",
    "#columns_to_check = ['CountGT_case_All', 'CountGT_control_All', 'CountGT_control_nonMiss', 'CountGT_case_nonMiss']\n",
    "\n",
    "#for col in columns_to_check:\n",
    "#    output_step2[col] = output_step2[col].apply(lambda x: len(x) if isinstance(x, list) else x)\n",
    "\n",
    "array_chisquare=output_step2.loc[:,['CountGT_case_All','CountGT_case_nonMiss','CountGT_control_All','CountGT_control_nonMiss']]\n",
    "array_chisquare['CountGT_case_All_other']=array_chisquare['CountGT_case_nonMiss']-array_chisquare['CountGT_case_All']\n",
    "array_chisquare['CountGT_control_All_other']=array_chisquare['CountGT_control_nonMiss']-array_chisquare['CountGT_control_All']\n",
    "contingency_table=array_chisquare.loc[:,['CountGT_case_All','CountGT_case_All_other','CountGT_control_All','CountGT_control_All_other']]\n",
    "\n",
    "print(array_chisquare[['CountGT_case_All', 'CountGT_case_All_other', 'CountGT_control_All', 'CountGT_control_All_other']].min())\n",
    "\n",
    "def chi_square_test(row):\n",
    "    obs = [[row['CountGT_case_All'], row['CountGT_case_All_other']],\n",
    "          [row['CountGT_control_All'], row['CountGT_control_All_other']]]\n",
    "    chi2, p, dof, ex = chi2_contingency(obs)\n",
    "    return p\n",
    "\n",
    "p=contingency_table.apply(chi_square_test, axis=1)\n",
    "output_step2['P_Chi-Square']=p\n",
    "print(output_step2)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9af9174d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
